Add paragraph mode to String#each_line in mrblib

lopopolo opened the pull request:
#4511 demonstrated an infinite loop in `String#each_line` when  …
given an empty string separator. In MRI, an empty separator places
String#each_line in paragraph mode, where the String is separated on
successive runs of newlines. In paragraph mode, the String
`"abc\n\n\ndef\nxyz"` is split into `["abc\n\n\n", "def\nxyz"]`.

This commit makes the String#each_line implementation as close to
ruby/spec compliant as possible given the limitations of mruby core.
With this patch, the following specs fail for `String#each_line`:

- uses `$/` as the separator when none is given (can be fixed by
  aliasing and redefining the method to use $/ as the default value
  of separator in mruby-io)
- when no block is given returned Enumerator size should return nil
  (`Enumerator#size` is not supported on mruby)
- tries to convert the separator to a string using to_str (`String#to_str`
  is not implemented on mruby)

This patch has similar memory consumption compared to the prior
implementation and is takes 4x the time the prior implementation takes to
execute:

```console
/usr/bin/time -l ./bin/mruby -e '("aaa\n\nbbbbb\n\n\n\n\ncccc" * 100000).each_line("\n") { }';
```

cc @dearblue 

Fixes GH-4511.

shuujii commented:
The second argument seems to be unnecessary.

lopopolo commented:
@shuujii the second argument is specified by CRuby. I've included it for compatibility.

lopopolo commented:
My latest commit improves performance of the base case to be within a factor of 2 of the existing implementation.

shuujii commented:
Because the second argument isn't used, it is better to remove it or implement correctly, I think.

BTW, I think `unless block` is faster than `unless block_given?` (ref. ef3bbd2e).

dearblue commented:
@lopopolo, there are two things I noticed in order to optimize the speed.

  - Pass position and length instead of passing range to get substrings.  
    Avoid range objects being generated each time in the loop.
  - Assign a method call whose value does not change in a loop to a variable outside the loop.  
    Because `==` is also a method, branch by putting a boolean value into the variable.

lopopolo commented:
> Because the second argument isn't used, it is better to remove it or implement correctly, I think.

I've removed this in the latest patch.

lopopolo commented:
The latest implementation is faster than the existing implementation while remaining as memory efficient. To test performance I ran this command:

```sh
/usr/bin/time -l ./bin/mruby -e '1000.times { ("aaa\n\nbbbbb\n\n\n\n\ncccc" * 10000).each_line("\n") { } }';
```

## Existing mrblib implementation

```
mruby@master
       17.91 real        17.84 user         0.06 sys
  54132736  maximum resident set size
         0  average shared memory size
         0  average unshared data size
         0  average unshared stack size
     45916  page reclaims
         0  page faults
         0  swaps
         0  block input operations
         0  block output operations
         0  messages sent
         0  messages received
         0  signals received
         0  voluntary context switches
      1472  involuntary context switches
```

## This PR at `8ea45b8`

```
mruby@string-each-line-paragraph-mode-8ea45b86
       17.91 real        17.84 user         0.05 sys
  54140928  maximum resident set size
         0  average shared memory size
         0  average unshared data size
         0  average unshared stack size
     45915  page reclaims
         0  page faults
         0  swaps
         0  block input operations
         0  block output operations
         0  messages sent
         0  messages received
         0  signals received
         0  voluntary context switches
      1343  involuntary context switches
```

## This PR at `6b92acf3`

```
mruby@string-each-line-paragraph-mode-6b92acf3
       17.40 real        17.34 user         0.05 sys
  54132736  maximum resident set size
         0  average shared memory size
         0  average unshared data size
         0  average unshared stack size
     45914  page reclaims
         0  page faults
         0  swaps
         0  block input operations
         0  block output operations
         0  messages sent
         0  messages received
         0  signals received
         0  voluntary context switches
      1219  involuntary context switches
```

lopopolo commented:
@shuujii and @dearblue thank you for helping me iterate on this PR. I think I've addressed all of the feedback which resulted in an implementation that is faster than the existing one while still being mostly spec-compliant. 🎉

dearblue commented:
It's a great :+1:

I realized I could replace it with a common loop.

example:

```ruby
if separator.empty?
  separator = "\n\n"
  paragraph_mode = true
end

while (pointer = string.index(separator, start))
  pointer += sep_len
  if paragraph_mode
    pointer += 1 while string[pointer] == "\n"
  end
  if should_yield_subclass_instances
    ...snip...
```

I'm worried about whether it fits your spec and the speed down, but I think the mruby bytecode size will be smaller.

Becomes like the dispensing of small information, I'm sorry.

lopopolo commented:
@dearblue thanks for the optimization! That does end up cutting down the code size quite a bit and it does still pass ruby-spec. 🎉 

I think this patch is ready for final review.
